### ISSUE TITLE
glam dependency is not optional in spirv-std

### ISSUE THREAD

**Issue by marstaik** [Rust-GPU/rust-gpu#146]
The readme in `spirv-std` states:
```
Optionally, through the use of the "glam" feature, it includes some boilerplate trait implementations to make glam vector types compatible with these APIs.
```

However, there is no feature flag that actually controls the glam bindings. As such, on a clean dependency with no features specified, spirv-std requires `glam` as a dependency to build.

I would definitely like to see this optional, and then perhaps introduce an nalgebra helper as well in the future.

**Comment by Firestar99** [comment]
`glam` used to be an optional dependency. Back then the sampling textures function would return a vector type of your choosing, specified in the `Image!` macro, that could convert from an array of 4 floats (or less if you specified the exact image type). But by specifying the returned vector type in the `Image!` macro, it would specify it via a generic on `Image` making it incompatible with any other type wanting a different vector type to be returned.

Some time ago it was decided to remove this feature of being able to return whatever vector type and sample() will just return glam's Vec4 (or Vec3, Vec2, float ...). And the end user should just use `mint` to convert it to whatever math library they're using.

If we really wanted to make `glam` optional again, these functions would have to return `[f32; 4]` which would make libraries using glam and not using glam incompatible, be a breaking change and it just adds more complexity for a relatively small library.

**Comment by schell** [comment]
We could try something like returning `T: impl From<[f32; 4]>`, but that would have implications for type inference.

**Comment by marstaik** [comment]
What if the Sampler types definitions or inner members where config guarded entirely? Instead of `[f32;4]`, make the features like `glam` and `nalgebra` mutually exclusive, and config guard the types depending on the feature used

**Comment by Firestar99** [comment]
Let's say we were to make those features. Then if you'd made a library that implements some technique, you need to choose between the `glam` or `nalgebra` feature to be able to sample textures at all, and you will be incompatible with any other library using the other. That's effectively splitting the ecosystem in two camps. 
Then I'd rather force-feed `glam` to everyone and tell them to convert it to the vector library of their choice.

**Comment by marstaik** [comment]
What would be the effective difference if glam remains the default enabled feature though? Glam would still be force fed with an option to change vector libraries

**Comment by marstaik** [comment]
> About `num-traits` failing to compile: It seems like you have some (transient) dependency on `num-traits` that requires the `std` feature. Could run `cargo tree -i num-traits -e features` and try to see if there's some **other library** depending on the `std` or `default` feature? Unfortunately, you cannot specify `--target spirv-unknown-vulkan1.2` due to default cargo not knowing of this target. So if you're running on current master, you will see that `spirv-std` depending on the default feature of `num-traits` _on CPU targets_. But when compiling for `spirv*` it _should_ instead disable default features and not actually depend on `default` / `std`.
> 
> For example, `nalgebra` with `default` depends on `nalgebra` with `std` depends on `num-traits` with `std`.

```
num-traits v0.2.19
├── approx v0.5.1
│   ├── nalgebra v0.33.2
│   │   ├── nalgebra feature "bytemuck"
│   │   │   ├── shaders v0.0.0 (/home/marios/proj/hestia_vk/packages/shaders)
│   │   │   │   └── shaders feature "default" (command-line)
│   │   │   └── vulkan v0.0.0 (/home/marios/proj/hestia_vk/packages/vulkan)
│   │   │       └── vulkan feature "default" (command-line)
│   │   │           └── hestia v0.0.0 (/home/marios/proj/hestia_vk/packages/hestia)
│   │   │               └── hestia feature "default" (command-line)
│   │   ├── nalgebra feature "macros"
│   │   │   ├── shaders v0.0.0 (/home/marios/proj/hestia_vk/packages/shaders) (*)
│   │   │   └── vulkan v0.0.0 (/home/marios/proj/hestia_vk/packages/vulkan) (*)
│   │   └── nalgebra feature "nalgebra-macros"
│   │       └── nalgebra feature "macros" (*)
│   └── simba v0.9.0
│       └── nalgebra v0.33.2 (*)
├── nalgebra v0.33.2 (*)
└── simba v0.9.0 (*)
├── num-traits feature "default"
│   ├── rustc_codegen_spirv v0.9.0 (https://github.com/Rust-GPU/rust-gpu.git#d5d347b0)
│   │   └── spirv-builder v0.9.0 (https://github.com/Rust-GPU/rust-gpu.git#d5d347b0)
│   │       ├── spirv-builder feature "default"
│   │       │   [build-dependencies]
│   │       │   └── shaders v0.0.0 (/home/marios/proj/hestia_vk/packages/shaders) (*)
│   │       └── spirv-builder feature "use-compiled-tools"
│   │           └── spirv-builder feature "default" (*)
│   │   └── rustc_codegen_spirv feature "use-compiled-tools"
│   │       └── spirv-builder feature "use-compiled-tools" (*)
│   └── spirv-std v0.9.0 (https://github.com/Rust-GPU/rust-gpu.git#d5d347b0)
│       └── spirv-std feature "default"
│           └── shaders v0.0.0 (/home/marios/proj/hestia_vk/packages/shaders) (*)
├── num-traits feature "i128"
│   ├── num-complex v0.4.6
│   │   ├── nalgebra v0.33.2 (*)
│   │   └── simba v0.9.0 (*)
│   ├── num-integer v0.1.46
│   │   └── num-integer feature "i128"
│   │       └── num-rational v0.4.2
│   │           └── nalgebra v0.33.2 (*)
│   └── num-rational v0.4.2 (*)
└── num-traits feature "std"
    └── num-traits feature "default" (*)
```
    
Seems like the only one requiring num-traits "default", and thus "std" is rust-gpu itself (if I am reading that correctly)

**Comment by marstaik** [comment]
I've uploaded a minimum reproduction here: https://github.com/marstaik/rust-gpu-repro

**Comment by LegNeato** [comment]
Perhaps if we stick to the `mint` traits the compiler could swap out the actual implementation at compile time? Similar to how when you open a file the compiler does different work on different platforms.

I could see a world where certain cards need certain specialized representations, especially TPUs and NPUs.

**Comment by Firestar99** [comment]
Found your issue: Your declared the resolver to be version 1 but must be version 2 for rust-gpu to work properly. Version 1 likes to "merge" features a bit too aggressively and is causing `num-traits` to get the `std` feature, even though it shouldn't.

I've made a [repo](https://github.com/Firestar99/rust-gpu-vulkano-example) detailing exactly how to setup rust-gpu with vulkano, feel free to take a look, even if you're graphics API is not vulkano it should be similar enough. We've already discussed internally that we need such an example repo for each major graphics API (wgpu, ash) and hopefully we'll get them ready before the next release.

**Comment by marstaik** [comment]
Thanks @Firestar99 , in addition to those two steps, I had to look at the example to enable some things like 
```rust
.multimodule(true)
.spirv_metadata(SpirvMetadata::NameVariables)
.print_metadata(MetadataPrintout::DependencyOnly)
.build()?;
```

despite not using vulkano, (I am using ash directly). But It does all seem to build now.

Back to the original topic, @LegNeato , using the mint trait would allow you to template/generic the Samplers by your own Vector type, correct?

So in user code we would probably have to do a type declaration like so:
``` rust
// nalegbra
pub type SamplerNa = SamplerImpl<Vector3<f32>>;

pub type SamplerGlam = SamplerImpl<Vec3>;
```

If that works, I don't mind putting some time into getting this working.

**Comment by Firestar99** [comment]
I knew there was another issue I found, but forgot when I started writing the list. All shader entry points must be accessible as `pub`, aka. both the entry point as well as the entire mod path all need to be `pub`. In your example your `mod example` is not :D

**Comment by LegNeato** [comment]
The `pub` stuff is https://github.com/EmbarkStudios/rust-gpu/issues/1140

**Comment by LegNeato** [comment]
@marstaik yeah, that's my thinking. I'm not experienced enough with this area to know if that's a good idea or not.
